### ISSUE TITLE
Consider supporting generic vulnerability scanners

### ISSUE THREAD

**Issue by nscuro** [DependencyTrack/dependency-track#1677]
We currently support multiple sources of vulnerability intelligence, among them the NVD, OSS Index, GHSA and VulnDB. In some cases, we perform the actual vulnerability scanning (e.g. NVD, GHSA), in others this is done by the sources themselves (e.g. OSS Index).

There are more great scanning tools out there, like Aqua's [Trivy](https://github.com/aquasecurity/trivy) or Anchore's [Grype](https://github.com/anchore/grype), that we could also integrate with. Trivy will soon support scanning of CycloneDX SBOMs (https://github.com/aquasecurity/trivy/pull/2203), and it already supports VEX output (https://github.com/aquasecurity/trivy/pull/1832). 

Given most tools are intended to be invoked via CLI, integration with DT may require a thin web API layer that exposes the functionality via REST or gRPC. 

Instead of hardcoding integrations in DT, we could allow for configuration of generic scanners that accept CycloneDX SBOMs and return CycloneDX VEX documents. Users could then decide to only rely on Trivy, or any other arbitrary combination of scanners.

Supporting tools like Trivy may also be beneficial for environments that are not supposed to have direct internet access (https://aquasecurity.github.io/trivy/v0.28.1/docs/advanced/air-gap/).

**Comment by nscuro** [comment]
More generally, this can be an architectural opportunity to decouple vulnerability detection from the API server. Even existing scanners like OSS Index could be decoupled in this way.

**Comment by tokcum** [comment]
I'm planning to use Trivy to generate SBOMs. Trivy can add the results of its own security vulnerability scan to the CycloneDX formated SBOM. However, when I import this SBOM file into DT only the components are imported. Vulns detected by Trivy don't make it into DT.

That's very unfortunate, as Trivy detected more vulns than DT. It would be great to see this happen. What does it take to push this issue forward?

**Comment by stevespringett** [comment]
@nscuro We need to think of a few different scenarios:

1) Import BOMs with embedded VDR (what this ticket is describing)
2) Import VDR BOMs (without component inventory)
3) Import Bill of Vulnerabilities

IMO, I think 1 and 3 will likely be the most used.

**Comment by Stamka** [comment]
> I'm planning to use Trivy to generate SBOMs. Trivy can add the results of its own security vulnerability scan to the CycloneDX formated SBOM. However, when I import this SBOM file into DT only the components are imported. Vulns detected by Trivy don't make it into DT.
> 
> That's very unfortunate, as Trivy detected more vulns than DT. It would be great to see this happen. What does it take to push this issue forward?

Do you solve this problem?

**Comment by ashutoshvimal** [comment]
any update?

**Comment by xico42** [comment]
any update on this?

It is very unfotunate that DT cannot read vulnerabilities from trivy generated SBOM :/

**Comment by iamrahul127** [comment]
Hi @stevespringett / @nscuro,
I see that this ticket is closed but I am still hopeful. Is there any possibility that DT can provide functionality to process vulnerabilities detected by trivy? 
My reflection is it will be helpful as DT not able to detect same issues for some reason.  When I upload a BOM generated by trivy DT could only identify 11 issues where as trivy reported 68 issues.

**Comment by nscuro** [comment]
> I see that this ticket is closed but I am still hopeful.

The issue is still open and implementation is definitely planned.

> When I upload a BOM generated by trivy DT could only identify 11 issues where as trivy reported 68 issues.

While this is a separate issue, are you able to share the BOM in question so we can investigate why DT doesn't find as many vulnerabilities?

**Comment by iamrahul127** [comment]
> The issue is still open and implementation is definitely planned.

Do you have any tentative release date?



> While this is a separate issue, are you able to share the BOM in question so we can investigate why DT doesn't find as many vulnerabilities?

I think [783](https://github.com/DependencyTrack/dependency-track/issues/783) is either same or related. BOM I have used is [myserver.web.txt](https://github.com/DependencyTrack/dependency-track/files/12924427/myserver.web.txt). Vulnerabilities from 'vulnerabilities' section of BOM are not processed even though they are present in vulnerabilities DB.
BOM screenshot for quick reference.

![image](https://github.com/DependencyTrack/dependency-track/assets/24877322/d2a77e6f-bdad-4095-981c-09de32bb8d4f)

Audit tab of DT show following.
![image](https://github.com/DependencyTrack/dependency-track/assets/24877322/8836a518-f384-4398-a0a7-83d16828c5cf)

**DT Configurations:**

- Installation & Version: AKS & 4.8.2
- Analysers enabled: Internal & Sonatype OSS Index
- Internal analyzer: Fuzzy CPE matching is disabled.
- Vulnerabilities sources: NVD & Github Advisories

Let me know if you need any more details.

**Comment by apsdts** [comment]
While trivy support is highly appreciated - I will try it soon - what about the initial question? 

We can already generate SBOMs with grype that include the results, but unfortunately dependency track does not consider them (which would be ok, if depednecy track would find the same issues like grype, which is not the case) .

**Comment by nscuro** [comment]
@apsdts Yes, what's described in this issue is still planned and #3251 is *not* a replacement for that.

**Comment by iamrahul127** [comment]
Apologies to trouble you again. When can we expect this PR? Are you targeting any specific release?

**Comment by iamrahul127** [comment]
> @apsdts Yes, what's described in this issue is still planned and #3251 is _not_ a replacement for that.

Our current plan is to run trivy in pipeline independently and upload SBOM generated to DT for further analysis. Trying to understand if we shouldn't be taking this approach and wait for #3251 and use integrated analyzer instead. Please help.

**Comment by fnxpt** [comment]
@iamrahul127 my 2 cents in this... 
To me it looks like a completely different issue, one is having the possibility of importing the vulnerabilities using a file like a vex, the other one is having an additional analyser.
If you run a pipeline and import a vex file you will be able to add the vulnerabilities but even if the analyser runs it wont be able to detect them.
Ideally we should have both, I was on vacation last 2 weeks I will try to find some time to see if I can add support for vex import soon

**Comment by cartermitchellLM** [comment]
Any update as to whether generic scanners are still planned or if it has been de-prioritized?

**Comment by nscuro** [comment]
@cartermitchellLM Trivy support has already been added. In v5 the analyzer API has further been abstracted such that analyzers consume CycloneDX BOMs and return results in CycloneDX VDR format: https://github.com/DependencyTrack/hyades-apiserver/blob/8419b5ddbffadce9ccb95ddcd3b684e7402a11b4/vuln-analysis/api/src/main/java/org/dependencytrack/vulnanalysis/api/VulnAnalyzer.java#L37-L132

This is already built on a plugin API which we may make available to external authors later. In that case, the community could create their own analyzers without modifying the core platform.

Alternatively, we just implement a generic analyzer with this API, ship it per default, and users can then point it to a HTTP / gRPC / whatever endpoint of their choice. We'd just need to define a proper API contract there.

In any case, contributions are welcome.

**Comment by cartermitchellLM** [comment]
Is there some documentation somewhere that details what you mean by "plugin API"?

**Comment by nscuro** [comment]
I have added some high-level docs on it here: https://dependencytrack.github.io/docs/next/concepts/architecture/design/extensibility/

The corresponding code is here: https://github.com/DependencyTrack/hyades-apiserver/tree/main/plugin

And for reference, the code of the OSS Index analyzer implemented on top of that mechanism: https://github.com/DependencyTrack/hyades-apiserver/tree/main/vuln-analysis/oss-index

That aside, there is also the possibility of us ingesting VDR documents (i.e., BOMs that already include vulnerability data), as mentioned earlier in this issue.

Which solution is the best way forward depends on your requirements:

1. Do you want DT to invoke the external analyzer?
    * Would be required if you want the analyzer to be considered not only for new BOM uploads but also for the scheduled (~daily) re-analysis.
    * Needs a new analyzer implementation as outlined in my previous comment.
2. Do you just want DT to store results an external tool already found?
    * In that case you accept that results become stale until you upload another VDR file.
    * Needs extension of the BOM ingestion mechanism to also consider vulnerability data.

(1) is what I had in mind when I created this issue. Support for (2) is already on the roadmap, but is insufficient for use-cases that don't involve high-frequency BOM uploads to keep data current.

So, the question is: What do you *need* from an "external analyzer" feature?

**Comment by mterron** [comment]
@nscuro I personally would love to have Option 1 as it seems like a natural fit for DepTrack. The api could be fully webhook driven and generic BOM in -> VDR out with all the scanner integration mess delegated to the Analyzer frontend. Even the Analyzer API could be generic in that all it needs to do is call into a script that drives the scanner.
```
DepTrack          Analyzer API            Analyzer
    |                    |                    |
    |---- POST BOM ----->|                    |
    |   + result URL     |                    |
    |<----- 200 OK ------|                    |
    |                    |                    |
    |                    |                    |
    |                    |--- Start scan  --->|
    |                    |<----- Results -----|
    |                    |                    |
    |                    |                    |
    |<---- POST VDR -----|                    |
    |   to result URL    |                    |
    |----- 200 OK ------>|                    |
```

**Comment by mterron** [comment]
Is this still on the table @nscuro ? Seems it'd be pretty simple to slightly modify the oss-index analyser to transform it into a "generic" analyzer. I presume we'd like the ability to configure multiple "generic" analysers but that shouldn't be too hard.

Happy to give it a go and create an initial PR
